File: packages/app/src/services/software/definitions/revolt.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory } from 'shared';
import { GithubReleasesScraper } from '../githubReleasesScraper';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'revolt';
	public name = 'Revolt';
	public category = [SoftwareCategory.Social];
	public downloadName = 'Revolt-Setup.exe';
	public shouldCacheUrl = true;
	public icon = 'revolt.png';
	public homepage = 'https://revolt.chat';

	public async resolveDownloadUrl() {
		const owner   = 'revoltchat';
		const repo    = 'desktop';
		const release = await GithubReleasesScraper.getRelease(owner, repo, r => r.endsWith('.exe'));
		if (release.isErr()) {
			return err(release.error);
		}

		return ok(release.value);
	}
}

File: src/lib/software/revolt.py
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from src.lib.github_release_scraper import GithubReleaseScraper

class Revolt(BaseSoftware):
    def __init__(self):
        super().__init__()

        self._gh = GithubReleaseScraper('revoltchat', 'desktop', self)
        self._gh.releases_scraped.connect(self._on_releases_scraped)

        self.key = 'revolt'
        self.name = 'Revolt'
        self.category = [SoftwareCategory.Social]
        self.download_name = 'Revolt-Setup.exe'
        self.should_cache_url = True
        self.icon = 'revolt.png'
        self.homepage = 'https://revolt.chat'

    @Slot(list)
    def _on_releases_scraped(self, releases: list[str]):
        asset = next((release for release in releases if release.endswith('.exe')), None)
        if asset:
            self.url_resolved.emit(asset)
        else:
            self.url_resolve_error.emit(self.ResolveError.GitHubAssetNotFoundError)

    def resolve_download_url(self):
        self._gh.get_repo_releases()
